W-22143873-websocket-limit-connection (#1052)
mulesoft/docs-gateway: gateway/1.13/modules/ROOT/pages/policies-included-websocket-connection-limit.adoc

= WebSocket Connection Limit Policy
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images
:keywords: websocket, connection limit, api gateway, flex gateway, gateway, policy

[width="100%", cols="5,15"]
|===
>s| Policy Name | WebSocket Connection Limit

> >s| Summary | Limits the number of simultaneous WebSocket connections that each API instance allows. Flex Gateway rejects new upgrade requests after reaching this limit.

Reviewer: ```suggestion
>s| Summary | Limits the number of simultaneous WebSocket connections that each API instance allows
```

>s| Category | Quality of Service
>s| First Flex Gateway version available | v1.13.0
.1+>.^s| Returned status codes

> | 429 - Flex Gateway reached the maximum number of concurrent WebSocket upgrades for the policy. The response includes the plain text `Too many WebSocket connections`.

Reviewer: ```suggestion
| 429 - `Too many WebSocket connections`
```

|===

== Summary

> The WebSocket Connection Limit policy tracks the number of active WebSocket connections for a specific API instance. Flex Gateway identifies these requests by the `Upgrade: websocket` header and ignores standard HTTP traffic. To ensure accuracy across distributed environments, Flex Gateway stores the connection counter in shared data so that all gateway replicas enforce a single, unified limit.

Reviewer: ```suggestion
The WebSocket Connection Limit policy tracks the number of active WebSocket connections for a specific API instance. Flex Gateway identifies these requests by the `Upgrade: websocket` header and ignores standard HTTP traffic. The WebSocket connection limit is distributed across Flex replicas.
```

When a WebSocket upgrade request arrives, Flex Gateway increments the counter to reserve capacity before forwarding the handshake to the upstream service. If the upstream service doesn't return a `101 Switching Protocols` response, Flex Gateway releases the reservation. After a successful WebSocket connection closes, Flex Gateway decrements the counter.

> If the counter reaches the `maximumConnections` limit, Flex Gateway rejects the upgrade request and returns a 429 error with the plain-text message `Too many WebSocket connections`.

Reviewer: ```suggestion
If the counter reaches the `maximumConnections` limit, Flex Gateway rejects the upgrade request with a `429` error code.
```

== Configuring Policy Parameters

include::partial$policy-title-headers.adoc[tag=configFile]

----
- policyRef:
    name: websocket-connection-limit-flex
  config:
    maximumConnections: <integer> // REQUIRED, minimum 1
----

[%header%autowidth.spread,cols="a,a,a,a"]
|===

| Parameter | Required | Default Value | Description

| `maximumConnections`
| Yes

> | —

Reviewer: Would these not have the same defualts?

| The maximum number of simultaneous WebSocket connections allowed. The value must be an integer of 1 or greater.

|===

==== Resource Configuration Example

[source,yaml]
----
- policyRef:
    name: websocket-connection-limit-flex
  config:
    maximumConnections: 500
----

include::partial$policy-title-headers.adoc[tag=ui]

[%header%autowidth.spread,cols="a,a,a,a"]
|===
s|Field s|Description s|Default s|Required

| Maximum Connections
| Maximum number of concurrent WebSocket connections allowed for the API instance.
| `100` | Yes
|===

> [NOTE]

Reviewer: Remove this note
